contracts/sdk/IpcContract.sol

```diff
@@ -0,0 +1,151 @@
+    // hasn't received an ACK.
+    mapping(bytes32 => IpcEnvelope) public inFlightMsgs;
+    // auxiliary mapping to track the height at which certain
+    // in-flight messages were received.
+    mapping(uint256 => bytes32[]) public inFlightMsgHeights;
+    // last height where there messages were pruned
+    uint256 private lastGcBlock;
+
+    // The adderss of the gateway in the subnet.
+    address public gatewayAddr;
+    /// Number of blocks that the contract will wait for a receipt
+    /// before garbage collecting the in-flight message.
+    uint256 public timeoutBlocks;
+
+    constructor(address gatewayAddr_, uint256 timeoutBlocks_) {
+        gatewayAddr = gatewayAddr_;
+        timeoutBlocks = timeoutBlocks_;
+    }
+
+    error CallerIsNotGateway();
+    error IpcMsgCallFailed();
+    error UnsupportedMsgKind();
+    error UnrecognizedReceipt();
+
+    /// @notice Entrypoint for IPC-enabled contracts. This function is always called by
+    /// the gateway when a `Call` or `Receipt` cross-net messages is targeted to
+    /// a specific address in the subnet.
+    function IpcEntrypoint(IpcEnvelope calldata crossMsg) external returns (bytes memory) {
+        // trigger the garbage collection of in-flight messages
+        // in case any timeout has triggered already.
+        // timeout == 0 means that no in-flight expiraton is set for receipts.
+        if (timeoutBlocks != 0) {
+            inFlightGC();
+        }
+
+        // only the gateway address is allowed to deliver cross-net messages.
+        if (msg.sender != gatewayAddr) {
+            revert CallerIsNotGateway();
+        }
+
+        // internal dispatch of the cross-net message to the right method.
+        if (crossMsg.kind == IpcMsgKind.Call) {
+            IpcMsg memory callMsg = abi.decode(crossMsg.message, (IpcMsg));
+
+            (bool success, bytes memory ret) = address(this).delegatecall(
+                abi.encodeWithSelector(callMsg.method, callMsg.params)
+            );
+            if (!success) {
+                revert IpcMsgCallFailed();
+            }
+
+            return ret;
+        } else if (crossMsg.kind == IpcMsgKind.Receipt) {
+            ReceiptMsg memory ackMsg = abi.decode(crossMsg.message, (ReceiptMsg));
+
+            // get the original message and chekc if the receipt was generated
+            // by the target contract, and that we actually sent that cross-net message.
+            // If this is not the case, revert, as someone may be trying to do something
+            // sketchy.
+            IpcEnvelope memory originalMsg = inFlightMsgs[ackMsg.id];
+            if (
+                originalMsg.message.length == 0 ||
+                keccak256(abi.encode(crossMsg.from)) != keccak256(abi.encode(originalMsg.to))
+            ) {
+                revert UnrecognizedReceipt();
+            }
+            // process ACK message and remove from in-flight if it succeeds.
+            if (AckEntrypoint(ackMsg)) {
```

Comment:
So here we would have correctly correlated the response, but handling the result can fail. Even in that case, we should remove the entry from the correlation map because we won't receive a second response.